freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UsersStateDifference.java

```diff
@@ -113,6 +125,39 @@ public static UsersStateDifference forDeletedUser(String deletedUser, Collection
         return usersToAdd;
     }
 
+    public static ImmutableSet<String> calculateUsersWithCredentialsToUpdate(UmsUsersState umsState, UsersState ipaState,
+            boolean credentialsUpdateOptimizationEnabled) {
+        ImmutableSet<String> usersWithCredentialsToUpdate;
+
+        if (!credentialsUpdateOptimizationEnabled) {
+            usersWithCredentialsToUpdate = ImmutableSet.copyOf(umsState.getUsersState().getUsers().stream()
+                    .map(FmsUser::getName)
+                    .filter(username -> !FreeIpaChecks.IPA_PROTECTED_USERS.contains(username))
+                    .collect(Collectors.toSet()));
+        } else {
+            ImmutableSet.Builder<String> usersWithCredentialsToUpdateBuilder = ImmutableSet.builder();
+            umsState.getUsersState().getUsers().stream()
+                    .map(FmsUser::getName)
+                    .filter(username -> !FreeIpaChecks.IPA_PROTECTED_USERS.contains(username))
+                    .forEach(username -> {
+                        UserMetadata ipaUserMetadata = ipaState.getUserMetadataMap().get(username);
+                        if (ipaUserMetadata == null) {
+                            usersWithCredentialsToUpdateBuilder.add(username);
+                        } else {
+                            WorkloadCredential umsCredential = umsState.getUsersWorkloadCredentialMap().get(username);
+                            if (ipaUserMetadata.getWorkloadCredentialsVersion() < umsCredential.getVersion()) {
+                                usersWithCredentialsToUpdateBuilder.add(username);
+                            }
+                        }
+                    });
```

Comment:
this should be also move to a separate method

Reply:
Done.